packages/find-root/src/index.ts

```diff
@@ -31,50 +33,60 @@ async function hasWorkspacesConfiguredViaPkgJson(
   }
 }
 
-async function hasWorkspacesConfiguredViaLerna(directory: string) {
-  try {
-    let lernaJson = await fs.readJson(path.join(directory, "lerna.json"));
-    if (lernaJson.useWorkspaces !== true) {
-      return directory;
-    }
-  } catch (err) {
-    if (err.code !== "ENOENT") {
-      throw err;
-    }
-  }
-}
+export async function findRoot(cwd: string): Promise<MonorepoRoot> {
+  let monorepoRoot: MonorepoRoot | undefined;
+
+  await findUp(
+    async directory => {
+      return Promise.all(
+        defaultOrder.map(toolType => supportedTools[toolType]).map(async (tool: Tool): Promise<MonorepoRoot | undefined> => {
+          if (await tool.isMonorepoRoot(directory)) {
+            return {
+              tool: tool,
+              dir: directory
+            };
+          }
+        })
+      ).then(x => x.find(value => value)).then(result => {
+        if (result && !monorepoRoot) {
+          monorepoRoot = result;
+        }
 
-async function hasWorkspacesConfiguredViaPnpm(directory: string) {
-  // @ts-ignore
-  let pnpmWorkspacesFileExists = await fs.exists(
-    path.join(directory, "pnpm-workspace.yaml")
+        // the findUp tool expects us to return the current directory or undefined
+        return result ? result.dir : undefined;
+      });
+    },
+    { cwd, type: "directory" }
   );
-  if (pnpmWorkspacesFileExists) {
-    return directory;
-  }
-}
 
-export async function findRoot(cwd: string): Promise<string> {
+  if (monorepoRoot)
+    return monorepoRoot;
+
+  // No monorepo root was found for any supported tool, so instead we return... something??
+  // I'm not sure what the case for this "default to current root pkg if no monorepo found"
+  // behavior is, it feels like we should return undefined and the calling tool should know
+  // that there is no monorepo.
+
   let firstPkgJsonDirRef: { current: string | undefined } = {
     current: undefined
   };
   let dir = await findUp(
     directory => {
       return Promise.all([
-        hasWorkspacesConfiguredViaLerna(directory),
-        hasWorkspacesConfiguredViaPkgJson(directory, firstPkgJsonDirRef),
-        hasWorkspacesConfiguredViaPnpm(directory)
+        hasWorkspacesConfiguredViaPkgJson(directory, firstPkgJsonDirRef)
       ]).then(x => x.find(dir => dir));
     },
     { cwd, type: "directory" }
   );
+
   if (firstPkgJsonDirRef.current === undefined) {
     throw new NoPkgJsonFound(cwd);
   }
-  if (dir === undefined) {
-    return firstPkgJsonDirRef.current;
-  }
-  return dir;
+
+  return {
+    tool: NoneTool,
```

Comment:
You have wanted to add an argument to control what kind of tools can be returned by a particular call. I wonder how this thing fits into that... especially that you didn't include `'none'` in the `defaultOrder` thing.

Reply:
Right -- I implemented logic in `NoneTool` to "find" a package.json file, but that is sort of wrong.

I think because NoneTool (or SinglePackageTool?) is an exception to regular rules, we need to formalize that exception in one of two ways:

 - SinglePackageTool's isMonorepoRoot always returns false no matter what
 - findRoot always excludes SinglePackageTool from its search list if it is included

I think option 1 is probably the best option, because then I can add SinglePackageTool to the end of the default search order, it would never be found as a monorepo root, and then findRoot has the special logic at the end to return it if you are sitting in a regular package folder.

Reply:
Sounds good when skimming over this. Could you implement this change (if you didn't do it already)?